File: dart/dnd_character.dart
import 'dart:math';

class DndCharacter {
  late int strength;
  late int constitution;
  late int dexterity;
  late int intelligence;
  late int wisdom;
  late int charisma;
  late int hitpoints;

  static int modifier(int score) {
    return ((score - 10) / 2).floor();
  }

  static int ability() {
    Random rand = Random();
    int sum = 0;
    int mi = 10;
    for (int i = 0; i < 4; i++) {
      final r = rand.nextInt(6) + 1;
      sum += r;
      mi = min(mi, r);
    }
    return sum - mi;
  }

  DndCharacter.create() {
    strength = ability();
    constitution = ability();
    dexterity = ability();
    intelligence = ability();
    wisdom = ability();
    charisma = ability();
    hitpoints = 10 + modifier(constitution);
  }
}

File: c/dnd_character.c
#include "dnd_character.h"
#include <math.h>
#include <stdlib.h>
#include <time.h>

int ability(void) {
  int sum = 0;
  int min = 10;
  for (int i = 0; i < 4; i++) {
    int r = rand() % 6 + 1;
    sum += r;
    if (min > r) {
      min = r;
    }
  }

  return sum - min;
}

int modifier(int score) { return floor((score - 10) / 2.0); }

dnd_character_t make_dnd_character(void) {
  srand((unsigned int)time(NULL));
  dnd_character_t character = {
      .wisdom = ability(),
      .constitution = ability(),
      .intelligence = ability(),
      .dexterity = ability(),
      .strength = ability(),
      .charisma = ability(),
      .hitpoints = 10 + modifier(character.constitution),
  };
  return character;
}
